cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,492 @@
+    }
+    // trivial for non-nullable columns
+    else {
+      thread_valid_count = thread_value_count;
+      block_valid_count  = block_value_count;
+    }
+
+    // output offset
+    if (is_valid) {
+      int const dst_pos = (value_count + thread_value_count) - 1;
+      int const src_pos = (valid_count + thread_valid_count) - 1;
+      auto ix           = rolling_index<state_buf::nz_buf_size>(src_pos);
+      sb->nz_idx[rolling_index<state_buf::nz_buf_size>(src_pos)] = dst_pos;
+    }
+
+    // update stuff
+    value_count += block_value_count;
+    valid_count += block_valid_count;
+  }
+
+  if (!t) {
+    // update valid value count for decoding and total # of values we've processed
+    ni.valid_count       = valid_count;
+    ni.value_count       = value_count;
+    s->nz_count          = valid_count;
+    s->input_value_count = value_count;
+    s->input_row_count   = value_count;
+  }
+
+  return valid_count;
+}
+
+template <typename state_buf>
+__device__ inline void gpuDecodeValues(
+  page_state_s* s, state_buf* const sb, int start, int end, int t)
+{
+  constexpr int num_warps      = decode_block_size / 32;
+  constexpr int max_batch_size = num_warps * 32;
+
+  PageNestingDecodeInfo* nesting_info_base = s->nesting_info;
+  int const dtype                          = s->col.data_type & 7;
+
+  // decode values
+  int pos = start;
+  while (pos < end) {
+    int const batch_size = min(max_batch_size, end - pos);
+
+    int const target_pos = pos + batch_size;
+    int const src_pos    = pos + t;
+
+    // the position in the output column/buffer
+    auto nz_idx = sb->nz_idx[rolling_index<state_buf::nz_buf_size>(src_pos)];
+    int dst_pos = sb->nz_idx[rolling_index<state_buf::nz_buf_size>(src_pos)] - s->first_row;
+
+    // target_pos will always be properly bounded by num_rows, but dst_pos may be negative (values
+    // before first_row) in the flat hierarchy case.
+    if (src_pos < target_pos && dst_pos >= 0) {
+      // nesting level that is storing actual leaf values
+      int leaf_level_index = s->col.max_nesting_depth - 1;
+
+      uint32_t dtype_len = s->dtype_len;
+      void* dst =
+        nesting_info_base[leaf_level_index].data_out + static_cast<size_t>(dst_pos) * dtype_len;
+      if (dtype == BYTE_ARRAY) {
+        if (s->col.converted_type == DECIMAL) {
+          auto const [ptr, len]        = gpuGetStringData(s, sb, src_pos);
+          auto const decimal_precision = s->col.decimal_precision;
+          if (decimal_precision <= MAX_DECIMAL32_PRECISION) {
+            gpuOutputByteArrayAsInt(ptr, len, static_cast<int32_t*>(dst));
+          } else if (decimal_precision <= MAX_DECIMAL64_PRECISION) {
+            gpuOutputByteArrayAsInt(ptr, len, static_cast<int64_t*>(dst));
+          } else {
+            gpuOutputByteArrayAsInt(ptr, len, static_cast<__int128_t*>(dst));
+          }
+        } else {
+          gpuOutputString(s, sb, src_pos, dst);
+        }
```

Comment:
We should strip out the string stuff here.

Reply:
This came up offline, but I _think_ the `strings_to_categories` path still uses `gpuOutputString()`. Maybe rather than `else` it could be `else if (s->dtype_len == 4)` to make that clearer.

Reply:
yes but not from the fixed* kernels as we are not letting any strings through here.

Reply:
the output type isn't string, it's int32, but the physical (input) type is byte_array. `is_string_col()` will return false for this case, so I think it will still wind up in this kernel. guess we could try testing that and see.

Reply:
oh, sorry, I see it now...you're testing for byte_array in `kernel_mask_for_page()`. nevermind :sweat_smile: